src/Shared/BinarySerializer.cs

```diff
@@ -4,43 +4,79 @@
 //
 
 using System.IO;
-using System.Runtime.Serialization.Formatters.Binary;
+using System.Text;
+using System.Web.SessionState;
+using ProtoBuf;
 
 namespace Microsoft.Web.Redis
 {
+
     public class BinarySerializer : ISerializer
     {
+        private readonly byte redisNull = 1;
+        private readonly byte initializeItem = 2;
+        private readonly byte str = 3;
         public byte[] Serialize(object data)
         {
-            if (data == null)
-            {
-                data = new RedisNull();
-            }
-            var binaryFormatter = new BinaryFormatter();
             using (var memoryStream = new MemoryStream())
             {
-                binaryFormatter.Serialize(memoryStream, data);
-                byte[] objectDataAsStream = memoryStream.ToArray();
-                return objectDataAsStream;
+                if (data is null)
+                {
+                    return new byte[] { redisNull };
+                }
+                if (data is SessionStateActions.InitializeItem)
+                {
+                    return new byte[] { initializeItem };
+                }
+                if (data is string)
+                {
+                    memoryStream.WriteByte(str);
+                }
+
```

Comment:
two things here: a) you *must* write a header in all cases (i.e. even when not `string`), because **you don't know** what the first byte of `Serializer.Serialize` will be; well... ok, in reality it is always going to be at least `>= 8`, but: please please use a header byte anyway

b) if the data is `string`, why still use `Serializer.Serialize`? seems to me you could just use `Encoding.Utf8` in that case, and avoid a step (and a lot of library spin)

c) (yes, I know I said "two things"; counting is hard) your comments (off github) suggest that `byte[]` is also a common case; perhaps we should special-case `byte[]` in the same was that we special-case `string`?

d) (yes, I hear myself) this seems alloc heavy; what's the throughput here? this looks to be a `public` API which makes refactoring it to be pool-friendly unlikely, but: there might still be options - for example, we could introduce an overloaded API that *is* pool friendly, use that exclusively, and have the legacy API (that we don't use) offer back-compat via alloc; happy to discuss in detail as needed

Reply:
This API isn't meant to be fully public - just some common code used by the various ASP.NET Providers our team owns in this repo. So a breaking API change is a possibility. If a pool would still benefit us, could you elaborate on that? Are you thinking a pool of byte[]s?